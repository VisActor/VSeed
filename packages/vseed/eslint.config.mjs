import js from '@eslint/js'
import globals from 'globals'
import ts from 'typescript-eslint'
import tseslint from 'typescript-eslint'

export default [
  { languageOptions: { globals: globals.browser } },
  js.configs.recommended,
  ...ts.configs.recommended,
  { ignores: ['dist/'] },

  ...tseslint.configs.recommendedTypeChecked,
  {
    files: ['src/**/*.ts', 'tests/**/*.ts'],
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.dirname,
      },
    },
    rules: {
      '@typescript-eslint/consistent-type-imports': 'error',
      '@typescript-eslint/consistent-type-exports': 'error',
      '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    },
  },
]
