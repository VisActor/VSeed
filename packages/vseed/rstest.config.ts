import { defineConfig } from '@rstest/core'

export default defineConfig({
  setupFiles: ['./rstest.setup.ts'],
  globals: true,
  testEnvironment: 'jsdom',
})
