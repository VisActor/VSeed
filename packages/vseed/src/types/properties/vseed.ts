import z from 'zod'
import type { ChartType } from './chartType'
import { zChartType } from './chartType'
import type { Dataset } from './dataset'
import { zDataset } from './dataset'
import type { Dimensions } from './dimensions'
import { zDimensions } from './dimensions'
import type { Measures } from './measures'
import { zMeasures } from './measures'

export const zVSeed = z.object({
  chartType: zChartType,
  dataset: zDataset,
  dimensions: zDimensions,
  measures: zMeasures,
})

export type VSeed = {
  chartType: ChartType
  dataset: Dataset
  dimensions: Dimensions
  measures: Measures
}

export const VSeedJSONSchema = z.toJSONSchema(zVSeed)
