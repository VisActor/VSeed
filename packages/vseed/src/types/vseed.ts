import { z } from 'zod'

import { zBar, type Bar } from './chartType/bar'
import { zBarParallel, type BarParallel } from './chartType/barParallel'
import { zColumn, type Column } from './chartType/column'
import { zColumnParallel, type ColumnParallel } from './chartType/columnParallel'
import { zColumnPercent, type ColumnPercent } from './chartType/columnPercent'
import { zLine, type Line } from './chartType/line'
import type { Table } from './chartType/table'
import type { Rose } from './chartType/rose'
import { zBarPercent, type BarPercent } from './chartType/barPercent'
import type { PivotTable } from './chartType/pivotTable'
import { zPie, type Pie } from './chartType/pie'
import type { Donut } from './chartType/donut'
import type { DualAxis } from './chartType/dualAxis'
import { zArea, type Area } from './chartType/area'
import { zAreaPercent, type AreaPercent } from './chartType/areaPercent'

export type VSeed =
  | Table
  | PivotTable
  | Line
  | Column
  | ColumnParallel
  | ColumnPercent
  | Bar
  | BarParallel
  | BarPercent
  | Area
  | AreaPercent
  | DualAxis
  | Rose
  | Pie
  | Donut

export const zVSeed = z.discriminatedUnion('chartType', [
  zLine,
  zColumn,
  zColumnParallel,
  zColumnPercent,
  zBar,
  zBarParallel,
  zBarPercent,
  zArea,
  zAreaPercent,
  zPie,
  // zRose,
  // zDualAxis,
  // zDonut,
])
