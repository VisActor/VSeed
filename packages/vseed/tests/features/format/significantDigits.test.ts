import type { VSeed } from '@visactor/vseed'
import { Builder, registerAll } from '@visactor/vseed'
import vseed from './significantDigits.json'

test('significantDigits', () => {
  registerAll()
  const builder = Builder.from(vseed as VSeed)
  const advanced = builder.buildAdvanced()
  if (!advanced) {
    throw new Error('Failed to build advanced configuration')
  }
  const spec = builder.buildSpec(advanced)
  expect(advanced).toMatchSnapshot()
  expect(spec).toMatchSnapshot()
});
