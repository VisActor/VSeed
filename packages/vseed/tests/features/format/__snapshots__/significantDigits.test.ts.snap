// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`significantDigits 1`] = `
{
  "annotation": {},
  "chartType": "column",
  "config": {
    "column": {
      "backgroundColor": "transparent",
      "color": {
        "colorScheme": [
          "#8D72F6",
          "#5766EC",
          "#66A3FE",
          "#51D5E6",
          "#4EC0B3",
          "#F9DF90",
          "#F9AD71",
          "#ED8888",
          "#E9A0C3",
          "#D77DD3",
        ],
        "linearColorScheme": [
          "#C2CEFF",
          "#5766EC",
        ],
      },
      "crosshairRect": {
        "labelBackgroundColor": "#364159",
        "labelColor": "#ffffff",
        "labelVisible": true,
        "rectColor": "#3641594d",
        "visible": true,
      },
      "label": {
        "enable": true,
        "labelColorSmartInvert": false,
        "labelOverlap": true,
        "showValue": true,
        "showValuePercent": false,
        "wrap": true,
      },
      "legend": {
        "border": true,
        "enable": true,
        "labelColor": "#646A73",
        "labelFontSize": 12,
        "labelFontWeight": 400,
        "maxSize": 1,
        "position": "rt",
        "shapeType": "rectRound",
      },
      "pivotGrid": {
        "bodyFontColor": "#141414",
        "borderColor": "#e3e5eb",
        "headerBackgroundColor": "#f6f7f9",
        "headerFontColor": "#21252c",
        "hoverHeaderBackgroundColor": "#D9DDE4",
        "hoverHeaderInlineBackgroundColor": "#D9DDE455",
      },
      "stackCornerRadius": [
        4,
        4,
        0,
        0,
      ],
      "tooltip": {
        "enable": true,
      },
      "xAxis": {
        "grid": {
          "gridColor": "#36415926",
          "gridWidth": 0.5,
          "visible": false,
        },
        "label": {
          "labelAngle": 0,
          "labelColor": "#797B85",
          "labelFontSize": 12,
          "labelFontWeight": 400,
          "visible": true,
        },
        "labelAutoHide": true,
        "labelAutoHideGap": 4,
        "labelAutoLimit": true,
        "labelAutoLimitLength": 80,
        "labelAutoRotate": false,
        "labelAutoRotateAngleRange": [
          0,
          -45,
          -90,
        ],
        "line": {
          "lineColor": "#3641594d",
          "lineWidth": 1,
          "visible": true,
        },
        "tick": {
          "tickColor": "#3641594d",
          "tickInside": false,
          "tickSize": 4,
          "visible": false,
        },
        "title": {
          "titleColor": "#646A73",
          "titleFontSize": 12,
          "titleFontWeight": 400,
          "titleText": "",
          "visible": false,
        },
      },
      "yAxis": {
        "grid": {
          "gridColor": "#36415926",
          "gridWidth": 0.5,
          "visible": true,
        },
        "inverse": false,
        "label": {
          "labelAngle": 0,
          "labelColor": "#797B85",
          "labelFontSize": 12,
          "labelFontWeight": 400,
          "visible": true,
        },
        "line": {
          "lineColor": "#3641594d",
          "lineWidth": 1,
          "visible": false,
        },
        "nice": true,
        "tick": {
          "tickColor": "#3641594d",
          "tickInside": false,
          "tickSize": 4,
          "visible": false,
        },
        "title": {
          "titleColor": "#646A73",
          "titleFontSize": 12,
          "titleFontWeight": 400,
          "titleText": "",
          "visible": false,
        },
        "zero": true,
      },
    },
  },
  "customTheme": {
    "config": {
      "column": {
        "backgroundColor": "transparent",
        "color": {
          "colorScheme": [
            "#8D72F6",
            "#5766EC",
            "#66A3FE",
            "#51D5E6",
            "#4EC0B3",
            "#F9DF90",
            "#F9AD71",
            "#ED8888",
            "#E9A0C3",
            "#D77DD3",
          ],
          "linearColorScheme": [
            "#C2CEFF",
            "#5766EC",
          ],
        },
        "crosshairRect": {
          "labelBackgroundColor": "#364159",
          "labelColor": "#ffffff",
          "labelVisible": true,
          "rectColor": "#3641594d",
          "visible": true,
        },
        "label": {
          "enable": true,
          "labelColorSmartInvert": false,
          "labelOverlap": true,
          "showValue": true,
          "showValuePercent": false,
          "wrap": true,
        },
        "legend": {
          "border": true,
          "enable": true,
          "labelColor": "#646A73",
          "labelFontSize": 12,
          "labelFontWeight": 400,
          "maxSize": 1,
          "position": "rt",
          "shapeType": "rectRound",
        },
        "pivotGrid": {
          "bodyFontColor": "#141414",
          "borderColor": "#e3e5eb",
          "headerBackgroundColor": "#f6f7f9",
          "headerFontColor": "#21252c",
          "hoverHeaderBackgroundColor": "#D9DDE4",
          "hoverHeaderInlineBackgroundColor": "#D9DDE455",
        },
        "stackCornerRadius": [
          4,
          4,
          0,
          0,
        ],
        "tooltip": {
          "enable": true,
        },
        "xAxis": {
          "grid": {
            "gridColor": "#36415926",
            "gridWidth": 0.5,
            "visible": false,
          },
          "label": {
            "labelAngle": 0,
            "labelColor": "#797B85",
            "labelFontSize": 12,
            "labelFontWeight": 400,
            "visible": true,
          },
          "labelAutoHide": true,
          "labelAutoHideGap": 4,
          "labelAutoLimit": true,
          "labelAutoLimitLength": 80,
          "labelAutoRotate": false,
          "labelAutoRotateAngleRange": [
            0,
            -45,
            -90,
          ],
          "line": {
            "lineColor": "#3641594d",
            "lineWidth": 1,
            "visible": true,
          },
          "tick": {
            "tickColor": "#3641594d",
            "tickInside": false,
            "tickSize": 4,
            "visible": false,
          },
          "title": {
            "titleColor": "#646A73",
            "titleFontSize": 12,
            "titleFontWeight": 400,
            "titleText": "",
            "visible": false,
          },
        },
        "yAxis": {
          "grid": {
            "gridColor": "#36415926",
            "gridWidth": 0.5,
            "visible": true,
          },
          "inverse": false,
          "label": {
            "labelAngle": 0,
            "labelColor": "#797B85",
            "labelFontSize": 12,
            "labelFontWeight": 400,
            "visible": true,
          },
          "line": {
            "lineColor": "#3641594d",
            "lineWidth": 1,
            "visible": false,
          },
          "nice": true,
          "tick": {
            "tickColor": "#3641594d",
            "tickInside": false,
            "tickSize": 4,
            "visible": false,
          },
          "title": {
            "titleColor": "#646A73",
            "titleFontSize": 12,
            "titleFontWeight": 400,
            "titleText": "",
            "visible": false,
          },
          "zero": true,
        },
      },
    },
  },
  "dataset": [
    {
      "__Dim_ColorId__": "num0-num0",
      "__Dim_Color__": "num0",
      "__Dim_Detail__": "num0",
      "__Dim_X__": "num0",
      "__MeaId__": "num0",
      "__MeaName__": "num0",
      "__MeaValue__": 123.456,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num0": 123.456,
      "num10": 1000,
    },
    {
      "__Dim_ColorId__": "num1-num1",
      "__Dim_Color__": "num1",
      "__Dim_Detail__": "num1",
      "__Dim_X__": "num1",
      "__MeaId__": "num1",
      "__MeaName__": "num1",
      "__MeaValue__": 123.456,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num1": 123.456,
      "num10": 1000,
    },
    {
      "__Dim_ColorId__": "num2-num2",
      "__Dim_Color__": "num2",
      "__Dim_Detail__": "num2",
      "__Dim_X__": "num2",
      "__MeaId__": "num2",
      "__MeaName__": "num2",
      "__MeaValue__": 123.456,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num2": 123.456,
    },
    {
      "__Dim_ColorId__": "num3-num3",
      "__Dim_Color__": "num3",
      "__Dim_Detail__": "num3",
      "__Dim_X__": "num3",
      "__MeaId__": "num3",
      "__MeaName__": "num3",
      "__MeaValue__": 123.456,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num3": 123.456,
    },
    {
      "__Dim_ColorId__": "num4-num4",
      "__Dim_Color__": "num4",
      "__Dim_Detail__": "num4",
      "__Dim_X__": "num4",
      "__MeaId__": "num4",
      "__MeaName__": "num4",
      "__MeaValue__": 123.456,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num4": 123.456,
    },
    {
      "__Dim_ColorId__": "num5-num5",
      "__Dim_Color__": "num5",
      "__Dim_Detail__": "num5",
      "__Dim_X__": "num5",
      "__MeaId__": "num5",
      "__MeaName__": "num5",
      "__MeaValue__": 234.567,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num5": 234.567,
    },
    {
      "__Dim_ColorId__": "num6-num6",
      "__Dim_Color__": "num6",
      "__Dim_Detail__": "num6",
      "__Dim_X__": "num6",
      "__MeaId__": "num6",
      "__MeaName__": "num6",
      "__MeaValue__": 234.567,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num6": 234.567,
    },
    {
      "__Dim_ColorId__": "num7-num7",
      "__Dim_Color__": "num7",
      "__Dim_Detail__": "num7",
      "__Dim_X__": "num7",
      "__MeaId__": "num7",
      "__MeaName__": "num7",
      "__MeaValue__": 234.567,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num7": 234.567,
    },
    {
      "__Dim_ColorId__": "num8-num8",
      "__Dim_Color__": "num8",
      "__Dim_Detail__": "num8",
      "__Dim_X__": "num8",
      "__MeaId__": "num8",
      "__MeaName__": "num8",
      "__MeaValue__": 345.678,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num8": 345.678,
    },
    {
      "__Dim_ColorId__": "num9-num9",
      "__Dim_Color__": "num9",
      "__Dim_Detail__": "num9",
      "__Dim_X__": "num9",
      "__MeaId__": "num9",
      "__MeaName__": "num9",
      "__MeaValue__": undefined,
      "__OriginalData__": {
        "num0": 123.456,
        "num1": 123.456,
        "num10": 1000,
        "num2": 123.456,
        "num3": 123.456,
        "num4": 123.456,
        "num5": 234.567,
        "num6": 234.567,
        "num7": 234.567,
        "num8": 345.678,
      },
      "num10": 1000,
      "num9": undefined,
    },
  ],
  "datasetReshapeInfo": [
    {
      "foldInfo": {
        "foldMap": {
          "num0": "num0",
          "num1": "num1",
          "num2": "num2",
          "num3": "num3",
          "num4": "num4",
          "num5": "num5",
          "num6": "num6",
          "num7": "num7",
          "num8": "num8",
          "num9": "num9",
        },
        "measureId": "__MeaId__",
        "measureName": "__MeaName__",
        "measureValue": "__MeaValue__",
        "statistics": {
          "colorMax": -Infinity,
          "colorMin": Infinity,
          "count": 10,
          "max": NaN,
          "min": NaN,
          "sum": NaN,
        },
      },
      "id": "column",
      "index": 0,
      "unfoldInfo": {
        "colorIdMap": {
          "num0-num0": "num0",
          "num1-num1": "num1",
          "num2-num2": "num2",
          "num3-num3": "num3",
          "num4-num4": "num4",
          "num5-num5": "num5",
          "num6-num6": "num6",
          "num7-num7": "num7",
          "num8-num8": "num8",
          "num9-num9": "num9",
        },
        "colorItems": [
          "num0-num0",
          "num1-num1",
          "num2-num2",
          "num3-num3",
          "num4-num4",
          "num5-num5",
          "num6-num6",
          "num7-num7",
          "num8-num8",
          "num9-num9",
        ],
        "encodingAngle": "__Dim_Angle__",
        "encodingColor": "__Dim_Color__",
        "encodingColorId": "__Dim_ColorId__",
        "encodingDetail": "__Dim_Detail__",
        "encodingX": "__Dim_X__",
        "encodingY": "__Dim_Y__",
      },
    },
  ],
  "dimensions": [
    {
      "alias": "指标名称",
      "id": "__MeaName__",
    },
  ],
  "encoding": {
    "color": [
      "__MeaName__",
    ],
    "column": [],
    "detail": [
      "__MeaName__",
    ],
    "label": [],
    "row": [],
    "tooltip": [],
    "x": [
      "__MeaName__",
    ],
    "y": [
      "num0",
      "num1",
      "num2",
      "num3",
      "num4",
      "num5",
      "num6",
      "num7",
      "num8",
      "num9",
    ],
  },
  "locale": "zh-CN",
  "markStyle": {},
  "measures": [
    {
      "alias": "num0",
      "format": {
        "significantDigits": 0,
        "type": "number",
      },
      "id": "num0",
    },
    {
      "alias": "num1",
      "format": {
        "significantDigits": 1,
        "type": "number",
      },
      "id": "num1",
    },
    {
      "alias": "num2",
      "format": {
        "significantDigits": 2,
        "type": "number",
      },
      "id": "num2",
    },
    {
      "alias": "num3",
      "format": {
        "significantDigits": 3,
        "type": "number",
      },
      "id": "num3",
    },
    {
      "alias": "num4",
      "format": {
        "significantDigits": 4,
        "type": "number",
      },
      "id": "num4",
    },
    {
      "alias": "num5",
      "format": {
        "significantDigits": 5,
        "type": "number",
      },
      "id": "num5",
    },
    {
      "alias": "num6",
      "format": {
        "significantDigits": 6,
        "type": "number",
      },
      "id": "num6",
    },
    {
      "alias": "num7",
      "format": {
        "significantDigits": 7,
        "type": "number",
      },
      "id": "num7",
    },
    {
      "alias": "num8",
      "format": {
        "significantDigits": 8,
        "type": "number",
      },
      "id": "num8",
    },
    {
      "alias": "num9",
      "format": {
        "significantDigits": 9,
        "type": "number",
      },
      "id": "num9",
    },
  ],
}
`;

exports[`significantDigits 2`] = `
{
  "animation": true,
  "axes": [
    {
      "domainLine": {
        "style": {
          "lineWidth": 1,
          "stroke": "#3641594d",
        },
        "visible": true,
      },
      "grid": {
        "style": {
          "lineWidth": 0.5,
          "stroke": "#36415926",
        },
        "visible": false,
      },
      "hover": true,
      "label": {
        "autoHide": true,
        "autoHideMethod": "greedy",
        "autoHideSeparation": 4,
        "autoLimit": true,
        "autoRotate": false,
        "autoRotateAngle": [
          0,
          -45,
          -90,
        ],
        "flush": true,
        "lastVisible": true,
        "minGap": 4,
        "space": 8,
        "style": {
          "angle": 0,
          "fill": "#797B85",
          "fontSize": 12,
          "fontWeight": 400,
          "maxLineWidth": 80,
        },
        "visible": true,
      },
      "maxHeight": 140,
      "orient": "bottom",
      "paddingInner": [
        0.15,
        0.1,
      ],
      "paddingOuter": [
        0.075,
        0.1,
      ],
      "sampling": false,
      "tick": {
        "inside": false,
        "style": {
          "stroke": "#3641594d",
        },
        "tickSize": 4,
        "visible": false,
      },
      "title": {
        "style": {
          "fill": "#646A73",
          "fontSize": 12,
          "fontWeight": 400,
        },
        "text": "指标名称",
        "visible": false,
      },
      "type": "band",
      "visible": true,
    },
    {
      "base": 10,
      "domainLine": {
        "style": {
          "lineWidth": 1,
          "stroke": "#3641594d",
        },
        "visible": false,
      },
      "grid": {
        "style": {
          "lineWidth": 0.5,
          "stroke": "#36415926",
        },
        "visible": true,
      },
      "innerOffset": {
        "top": 7,
      },
      "inverse": false,
      "label": {
        "formatMethod": [Function],
        "style": {
          "angle": 0,
          "fill": "#797B85",
          "fontSize": 12,
          "fontWeight": 400,
        },
        "visible": true,
      },
      "max": undefined,
      "min": undefined,
      "nice": true,
      "orient": "left",
      "tick": {
        "inside": false,
        "style": {
          "stroke": "#3641594d",
        },
        "tickSize": 4,
        "visible": false,
      },
      "title": {
        "style": {
          "fill": "#646A73",
          "fontSize": 12,
          "fontWeight": 400,
        },
        "text": "num0 & num1 & num2 & num3 & num4 & num5 & num6 & num7 & num8 & num9",
        "visible": false,
      },
      "type": "linear",
      "visible": true,
      "zero": true,
    },
  ],
  "background": "transparent",
  "bar": {
    "state": {
      "hover": {
        "fillOpacity": 0.6,
      },
    },
    "style": {
      "lineWidth": 1,
    },
  },
  "color": {
    "domain": [
      "num0-num0",
      "num1-num1",
      "num2-num2",
      "num3-num3",
      "num4-num4",
      "num5-num5",
      "num6-num6",
      "num7-num7",
      "num8-num8",
      "num9-num9",
    ],
    "range": [
      "#8D72F6",
      "#5766EC",
      "#66A3FE",
      "#51D5E6",
      "#4EC0B3",
      "#F9DF90",
      "#F9AD71",
      "#ED8888",
      "#E9A0C3",
      "#D77DD3",
    ],
    "specified": undefined,
    "type": "ordinal",
  },
  "crosshair": {
    "xField": {
      "label": {
        "labelBackground": {
          "style": {
            "fill": "#364159",
          },
          "visible": true,
        },
        "style": {
          "fill": "#ffffff",
        },
        "visible": true,
      },
      "line": {
        "style": {
          "fill": "#3641594d",
          "lineWidth": 0,
          "opacity": 0.2,
        },
        "type": "rect",
      },
      "visible": true,
    },
  },
  "data": {
    "fields": {
      "__Dim_Angle__": {
        "sortIndex": 0,
      },
      "__Dim_ColorId__": {
        "sortIndex": 0,
      },
      "__Dim_X__": {
        "sortIndex": 0,
      },
    },
    "id": "column",
    "values": [
      {
        "__Dim_ColorId__": "num0-num0",
        "__Dim_Color__": "num0",
        "__Dim_Detail__": "num0",
        "__Dim_X__": "num0",
        "__MeaId__": "num0",
        "__MeaName__": "num0",
        "__MeaValue__": 123.456,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num0": 123.456,
        "num10": 1000,
      },
      {
        "__Dim_ColorId__": "num1-num1",
        "__Dim_Color__": "num1",
        "__Dim_Detail__": "num1",
        "__Dim_X__": "num1",
        "__MeaId__": "num1",
        "__MeaName__": "num1",
        "__MeaValue__": 123.456,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num1": 123.456,
        "num10": 1000,
      },
      {
        "__Dim_ColorId__": "num2-num2",
        "__Dim_Color__": "num2",
        "__Dim_Detail__": "num2",
        "__Dim_X__": "num2",
        "__MeaId__": "num2",
        "__MeaName__": "num2",
        "__MeaValue__": 123.456,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num2": 123.456,
      },
      {
        "__Dim_ColorId__": "num3-num3",
        "__Dim_Color__": "num3",
        "__Dim_Detail__": "num3",
        "__Dim_X__": "num3",
        "__MeaId__": "num3",
        "__MeaName__": "num3",
        "__MeaValue__": 123.456,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num3": 123.456,
      },
      {
        "__Dim_ColorId__": "num4-num4",
        "__Dim_Color__": "num4",
        "__Dim_Detail__": "num4",
        "__Dim_X__": "num4",
        "__MeaId__": "num4",
        "__MeaName__": "num4",
        "__MeaValue__": 123.456,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num4": 123.456,
      },
      {
        "__Dim_ColorId__": "num5-num5",
        "__Dim_Color__": "num5",
        "__Dim_Detail__": "num5",
        "__Dim_X__": "num5",
        "__MeaId__": "num5",
        "__MeaName__": "num5",
        "__MeaValue__": 234.567,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num5": 234.567,
      },
      {
        "__Dim_ColorId__": "num6-num6",
        "__Dim_Color__": "num6",
        "__Dim_Detail__": "num6",
        "__Dim_X__": "num6",
        "__MeaId__": "num6",
        "__MeaName__": "num6",
        "__MeaValue__": 234.567,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num6": 234.567,
      },
      {
        "__Dim_ColorId__": "num7-num7",
        "__Dim_Color__": "num7",
        "__Dim_Detail__": "num7",
        "__Dim_X__": "num7",
        "__MeaId__": "num7",
        "__MeaName__": "num7",
        "__MeaValue__": 234.567,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num7": 234.567,
      },
      {
        "__Dim_ColorId__": "num8-num8",
        "__Dim_Color__": "num8",
        "__Dim_Detail__": "num8",
        "__Dim_X__": "num8",
        "__MeaId__": "num8",
        "__MeaName__": "num8",
        "__MeaValue__": 345.678,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num8": 345.678,
      },
      {
        "__Dim_ColorId__": "num9-num9",
        "__Dim_Color__": "num9",
        "__Dim_Detail__": "num9",
        "__Dim_X__": "num9",
        "__MeaId__": "num9",
        "__MeaName__": "num9",
        "__MeaValue__": undefined,
        "__OriginalData__": {
          "num0": 123.456,
          "num1": 123.456,
          "num10": 1000,
          "num2": 123.456,
          "num3": 123.456,
          "num4": 123.456,
          "num5": 234.567,
          "num6": 234.567,
          "num7": 234.567,
          "num8": 345.678,
        },
        "num10": 1000,
        "num9": undefined,
      },
    ],
  },
  "direction": "vertical",
  "label": {
    "formatMethod": [Function],
    "overlap": {
      "clampForce": true,
      "hideOnHit": true,
    },
    "position": undefined,
    "smartInvert": false,
    "style": {
      "background": undefined,
      "fill": undefined,
      "fontSize": undefined,
      "fontWeight": undefined,
    },
    "visible": true,
  },
  "legends": {
    "autoPage": true,
    "data": [Function],
    "item": {
      "background": {
        "state": {
          "selectedHover": {
            "fill": "#646A73",
            "fillOpacity": 0.05,
          },
        },
      },
      "focus": true,
      "focusIconStyle": {
        "fill": "#646A73",
        "fontWeight": 400,
        "size": 14,
      },
      "label": {
        "formatMethod": [Function],
        "style": {
          "fill": "#646A73",
          "fontSize": 12,
          "fontWeight": 400,
        },
      },
      "maxWidth": "30%",
      "shape": {
        "space": 6,
        "style": {
          "size": 8,
          "symbolType": "rectRound",
        },
      },
    },
    "maxCol": 1,
    "maxRow": 1,
    "orient": "right",
    "padding": 0,
    "position": "start",
    "type": "discrete",
    "visible": true,
  },
  "padding": 0,
  "region": [
    {
      "clip": true,
    },
  ],
  "seriesField": "__Dim_ColorId__",
  "stackCornerRadius": [Function],
  "stackInverse": true,
  "tooltip": {
    "dimension": {
      "content": [
        {
          "hasShape": true,
          "key": [Function],
          "shapeType": "rectRound",
          "value": [Function],
          "visible": true,
        },
      ],
      "title": {
        "visible": true,
      },
    },
    "mark": {
      "content": [
        {
          "hasShape": true,
          "key": [Function],
          "shapeType": "rectRound",
          "value": [Function],
          "visible": true,
        },
      ],
      "title": {
        "visible": false,
      },
    },
    "visible": true,
  },
  "type": "bar",
  "xField": "__Dim_X__",
  "yField": "__MeaValue__",
}
`;

exports[`significantDigits 3`] = `
{
  "advancedPipeline": [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "colorIdMap": {
    "num0-num0": "num0",
    "num1-num1": "num1",
    "num2-num2": "num2",
    "num3-num3": "num3",
    "num4-num4": "num4",
    "num5-num5": "num5",
    "num6-num6": "num6",
    "num7-num7": "num7",
    "num8-num8": "num8",
    "num9-num9": "num9",
  },
  "colorItems": [
    {
      "alias": "num0",
      "id": "num0-num0",
    },
    {
      "alias": "num1",
      "id": "num1-num1",
    },
    {
      "alias": "num2",
      "id": "num2-num2",
    },
    {
      "alias": "num3",
      "id": "num3-num3",
    },
    {
      "alias": "num4",
      "id": "num4-num4",
    },
    {
      "alias": "num5",
      "id": "num5-num5",
    },
    {
      "alias": "num6",
      "id": "num6-num6",
    },
    {
      "alias": "num7",
      "id": "num7-num7",
    },
    {
      "alias": "num8",
      "id": "num8-num8",
    },
    {
      "alias": "num9",
      "id": "num9-num9",
    },
  ],
  "hasTheme": true,
  "specPipeline": [
    [Function],
  ],
  "themeMapKeys": [
    "light",
    "dark",
  ],
}
`;
